onmt/modules/Embeddings.py

```diff
@@ -0,0 +1,147 @@
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+from onmt.modules import BottleLinear
+from onmt.modules import aeq
+
+
+class Elementwise(nn.ModuleList):
+    """
+    A simple network container.
+    Parameters are a list of modules.
+    Inputs are a 3d Variable whose last dimension is the same length
+    as the list.
+    Outputs are the result of applying modules to inputs elementwise.
+    An optional merge parameter allows the outputs to be reduced to a
+    single Variable.
+    """
+
+    def __init__(self, merge=None, *args):
+        assert merge in [None, 'first', 'concat', 'sum', 'mlp']
+        self.merge = merge
+        super(Elementwise, self).__init__(*args)
+
+    def forward(self, input):
+        inputs = [feat.squeeze(2) for feat in input.split(1, dim=2)]
+        assert len(self) == len(inputs)
+        outputs = [f(x) for f, x in zip(self, inputs)]
+        if self.merge == 'first':
+            return outputs[0]
+        elif self.merge == 'concat' or self.merge == 'mlp':
+            return torch.cat(outputs, 2)
+        elif self.merge == 'sum':
+            return sum(outputs)
+        else:
+            return outputs
+
+
+class PositionalEncoding(nn.Module):
```

Comment:
Nice, I like that this is a separate class. It seems like people are also using LUT based positions, so this makes it easy to add that. 
Q: will `self.pe` work here with cuda too?

Reply:
Hmm, I thought it would because the `PositionalEncoding` instance is ultimately one of the submodules of the NMTModel. But now I'm not convinced. I think it may need to be registered as a buffer.